### ISSUE TITLE
Misbehaviour (again) while scaling parts with VARIANT

### ISSUE THREAD

**Issue by Lisias** [net-lisias-ksp/TweakScale#261]
When you change the variant of a scaled part that it's attached backwards, the part is terribly mispositioned.

On a wild guess, I think I'm using a global coordinate when I should had been used a local one.

![screenshot69](https://user-images.githubusercontent.com/64334/179405498-cc297b69-1391-4c0c-8abd-b5a349c420c8.png)

![screenshot71](https://user-images.githubusercontent.com/64334/179405505-d95a9e45-e0c3-43b7-a5e4-24c4389fbd29.png)

![screenshot72](https://user-images.githubusercontent.com/64334/179405512-fb42b6a8-d12b-410b-ac8e-639cd0251df2.png)

On KSP 1.7.3 this is happening too when reverting to VAB/SPH. Interesting… On 1.12.3 I have KSP-Recall installed, on 1.7.3 I don't.

Anyway, I think that by fixing the misbehaviour on the Scaling Engine for Variants I will solve the problem on both KSP versions.

**Comment by Lisias** [comment]
Unsurprisingly, the misbehaviour couldn't be reproduced on KSP 1.4.3. So the "Classic" scaling engine is working fine, it's something I let pass trough on the Variant scaling engine.

**Comment by Lisias** [comment]
The problem **was** reproduced on KSP 1.4.5.

So it is, indeed, a problem on the Variant Scaling Engine.

**Comment by Lisias** [comment]
> Unsurprisingly, the misbehaviour couldn't be reproduced on KSP 1.4.3. So the "Classic" scaling engine is working fine, it's something I let pass trough on the Variant scaling engine.

**I'M AN IDIOT**…. Ok, no news here. :)

I could reproduce the problem on KSP 1.4.3 too . I forgot there were already some variants on 1.4.3. (Alzheimer, here we go!)

![screenshot66](https://user-images.githubusercontent.com/64334/180208574-2d914789-a37a-41e1-97a6-7ade4f7d95d3.png)

**Comment by Lisias** [comment]
This zero in the problem on the original Variant Scaling Engine (Scale.PartDB.14x).

**Comment by Lisias** [comment]
Interesting…. This does not appears to be a problem on TweakScale after all.

Launching the test vessel directly into the Runway works fine.

***KRAP***. I'm innocent after all, this is that crapness from KSP Editor. Well, well… This started on KSP 1.4.3 after all…

![Uploading screenshot67.png…]()

**Comment by Lisias** [comment]
Well, we have **TWO** misbehaviours working in parallel.

One is on KSP's Editor, and this krapness is happening since 1.4.3 - the very first implementation for the PartModuleVariant. Damn, what a crappy implementation they did…

Loading it directory into Launch as the previous comment states always work fine.
![screenshot77](https://user-images.githubusercontent.com/64334/188302454-c93065a0-de00-4fb6-9a01-942e63358a8c.png)

This will be tacked down on https://github.com/net-lisias-ksp/KSP-Recall/issues/55

The second misbehaviour is related to changing the Variant of a part when it is attached "inverted". This one appears to be on me.

![screenshot79](https://user-images.githubusercontent.com/64334/188302467-bac838c8-7d29-47bc-b297-6d0aea36b369.png)
![screenshot80](https://user-images.githubusercontent.com/64334/188302471-6155bbd0-82ed-4b19-982d-3bec1f19d152.png)

Interesting to note that the code is working fine when the part is attached "with the proper orientation". Only when attaching it backwards the problem happens.

**Comment by Lisias** [comment]
The KSP's Editor misbehaviour is fixed by https://github.com/net-lisias-ksp/KSP-Recall/releases/tag/RELEASE%2F0.3.0.2

Now back to the problem that is really on my side. (damn, I wasted too much time on that crap)

**Comment by Lisias** [comment]
Well.. That's the deal.

On KSP 1.4.3 (or 1.4.2, the Brief), when `ModulePartVariants` became a thing Squad's broke the Editor (I'm guessing to workaround some flaw on the PartModule). Original stack attachments started to be overwritten `OnLoad` royally screwing up TweakScale when used on parts with Variants - at that time, only a few on Making History.

When 1.9.x came and I finally detected a misbehaviour on Editor, this happened because KSP Editor started to screw up the Surface Attachments too. This was detected and "fixed" on [Issue #110](https://github.com/net-lisias-ksp/TweakScale/issues/110). The fix, obviously, caused some collateral effects because it was built over a "gambiarra" made due the first Editor misbehaviour - destabilizing the chain of events that everybody needs to happen to co-exist.

TL;DR: almost all the [issues](https://github.com/net-lisias-ksp/TweakScale/search?p=2&q=variant) related to Variant happened due this initial misdiagnosing… (sigh).

Oh, well… better late then never.

**Comment by Lisias** [comment]
OUKEY. The behaviour depicted on his [comment](https://github.com/net-lisias-ksp/TweakScale/issues/261#issuecomment-1191397081) and the picture below:

![](https://user-images.githubusercontent.com/64334/180208574-2d914789-a37a-41e1-97a6-7ade4f7d95d3.png)

can't be reproduced on Beta.

However, by changing the Variant after scaling the part, the scaled "sinks" info it's parent!

1. Subject
* ![screenshot4](https://user-images.githubusercontent.com/64334/212561574-1abc99fd-9c23-4fec-b9fd-1b85ddb7a3df.png)
2. Changing the variant of a scaled part:
* ![screenshot5](https://user-images.githubusercontent.com/64334/212561595-51887f88-8824-4780-b23f-f29b5803ea3f.png)
3. Results:
* ![screenshot6](https://user-images.githubusercontent.com/64334/212561607-d7cc8b59-1a0d-43fb-8c49-cdfbdd909e64.png)
4. Doing it on the other part
* ![screenshot7](https://user-images.githubusercontent.com/64334/212561624-0f6a9298-cf14-4215-b226-a676f021d636.png)
5. Doing it again on the last part - the effect is cumulative!
* ![screenshot8](https://user-images.githubusercontent.com/64334/212561668-18e5b33f-dea6-4306-906b-64bf21a7aded.png)

**Comment by Lisias** [comment]
Oukey, zeroed on the problem.

I'm failing on fetching the correct AttachmentNode from the new PartVariant - or perhaps I'm doing it right, but I'm being overruled later in the chain?

**Comment by Lisias** [comment]
Ran out of time for this. But at least I understood where the problem is.

I'm postponing this again, and I will release 2.4.6.19 with what I have at hands now.

**Comment by Lisias** [comment]
Less is more (hopefully).

I decided to trust (or hope) ModulePartVariant is doing its job (what appears to be true - most of the time), and so just triggering a full Rescale when the Variant is changed.

Lets see what's happens - at least for the simpler Variants, this is working.

Implemented on commit https://github.com/net-lisias-ksp/TweakScale/commit/7732b7c97516e03ca5149faaf88e77f4a601aabe

**Comment by Lisias** [comment]
Oukey, the thing is confirmed to work fine on KSP 1.4.x.

![screenshot85](https://user-images.githubusercontent.com/64334/214252432-77e23050-5640-4f93-9bc0-cac20148da20.png)

**Comment by Lisias** [comment]
But then it misbehaves on 1.5.x - not a surprise, as I built the thing to be "stacked", one support built over the previous. Since I "broke" 1.4.x to make it "right" this time, it's expected that it broke 1.5.x in the process (and so on).

Interesting enough, the fix was just getting rid of the 1.4.x code!

![screenshot18](https://user-images.githubusercontent.com/64334/214254607-efa363a2-ede1-4cc4-94ce-e6c7d92a6d7e.png)

Fix on commit https://github.com/net-lisias-ksp/TweakScale/commit/cd885781825b54559ad35b12d88d916e45fee098

**Comment by Lisias** [comment]
KSP 1.7.x behave as expected.

**Comment by Lisias** [comment]
KSP 1.8.x also worked fine. Great! This concludes the 1.5.x scaler testings! Now working on the 1.9.x series (that goes until 1.12.x)

**Comment by Lisias** [comment]
And the KSP 1.9.x series also behaved!! **#HURRAY!!**

![screenshot63](https://user-images.githubusercontent.com/64334/214262557-65408652-9c38-49c8-bbd5-3749bb8d5880.png)

The terrible image quality is due the need to shove it into the ground on my rig, as Squad had decided to use 4K textures for everything, screwing up with everybody using cheaper GPUs. I will solve this problem later, on KSP-Recall or pehaps some other add'on?

**Comment by Lisias** [comment]
Oukey. 1.12.5 is working fine, now I can close this one.

This is a historical day. I **finally** tackled down one of the **worst** (if not the worst) problems I ever faced on KSP: supporting the `ModulePartVariant`.

The module itself is not bad, the problem is Editor. Since 1.4.x Editor is broken due a bad implementation on supporting this Module - the dudes just shoved some 'gambiarra' on the code that **royally** screwed **everything and everybody** that allowed changing things in Editor time. See KSP-Recall's issues and code related to `AttachedOnEditor` for more information.

This literally took me **years** to figure out - I initially misdiagnosed the problem to be 1.9.x related (just because `PartModuleVariant` started to support Attachment Points on the variants too on 1.9.x), and so I ended coding some shitty code on Scalers for 1.4.x to 1.8.x without being aware of the root problem.

Once I finally diagnose the mess properly, I had to virtually rewrite (to tell you the true, to **remove** code) support for scaling Variants since the very first implementation - because I was coding around a bug, not a feature, and this code caused a lot of collateral effects, probably on 3rd parties.

Oh well.. Better later than never, now I know and and fixed things properly. At least until KSP 1.13.0 (if ever) , when they may create new ways of screwing us over! :P

Closing this for good now.
